onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -330,10 +330,29 @@ Status CUDAExecutionProvider::OnRunStart() {
   auto& current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventCreate(&current_deferred_release_event, cudaEventDisableTiming));
   deferred_release_cpu_ptr_.emplace(current_deferred_release_event, DeferredReleaseCPUPtrs());
+
+  if (ConfiguredForGraphCapture() && !IsGraphCaptured()) {
+    if (FinishRegularRunsBeforeGraphCapture()) {
+      LOGS_DEFAULT(INFO) << "Capturing the CUDA Graph for this model";
+      CaptureBegin();
+    } else {
+      LOGS_DEFAULT(INFO) << "Run the " << regular_run_count_before_graph_capture_ << "th of the required " << DEFAULT_REGULAR_RUNS_BEFORE_CUDA_GRAPH_CAPTURE << " regular runs for cuda graph capture. ";
+    }
+  }
   return Status::OK();
 }
 
 Status CUDAExecutionProvider::OnRunEnd(bool sync_stream) {
+  if (ConfiguredForGraphCapture() && !IsGraphCaptured()) {
+    if (FinishRegularRunsBeforeGraphCapture()) {
+      CaptureEnd();
+      // CUDA work issued to a capturing stream doesn’t actually run on the GPU,
+      // so run the captured graph here to actually execute the work.
+      ORT_RETURN_IF_ERROR(graph_.Replay());
+    } else {
+      ++regular_run_count_before_graph_capture_;
```

Comment:
I just realized that `++regular_run_count_before_graph_capture_` doesn't seem to be thread-safe especially since Run() is supposed to safe to be called from different threads. But then again, triggering graph capture while one capture is going on will cause issues and we don't have a guard against that either.

I think `++regular_run_count_before_graph_capture_` can be moved into `FinishRegularRunsBeforeGraphCapture()` with a lock and maybe there should be a lock used in `CUDAGraph::CaptureBegin()` and `CaptureEnd()` and just return back if the graph capture has been done using another thread ?

Reply:
Good idea. Cuda graph with multiple threads are supported by leveraging `PerThreadContext`.